README.md
# wallet

Will be evolving over time. Currently it can pull market transaction for characters and corporations, and apply some filters.

## Usage

REPL usage only for now. Maybe I'll make a simple text UI as well :D. This is meant to be a library for use in other applications.


## Examples

Get totals for all transactions of type "85" (Bounty Prizes) in the time range (previous week)

```clojure
(eve.wallet/balance-info 
              {:keyID "<key id>"
               :vCode "<verification code>"}
              {:refTypeID (partial = "85")
               :date (fn [date]
                       (t/after? (f/parse date) (t/minus (t/now) (t/weeks 1))))})
```

## License

Copyright © 2017 Sean Henley

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-02-25
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-02-25
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/wallet/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/wallet/compare/0.1.0...0.1.1


project.clj
(defproject wallet "0.1.0-SNAPSHOT"
  :description "Collections of eve tools and helper functions"
  :url "http://example.com/FIXME"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [org.clojure/data.xml "0.2.0-alpha2"]
                 [clj-time "0.13.0"]
                 ;; Logging specific deps
                 [org.clojure/tools.logging "0.3.1"]
                 [org.slf4j/slf4j-log4j12 "1.7.1"]
                 [log4j/log4j "1.2.17" :exclusions [javax.mail/mail
                                                    javax.jms/jms
                                                    com.sun.jmdk/jmxtools
                                                    com.sun.jmx/jmxri]]]
  :main ^:skip-aot eve.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})


src/wallet/eve.clj
(ns wallet.eve
  "eve api related helpers"
  (:require [clojure.data.xml :as xml]
            [clojure.string :as str]
            [clojure.tools.logging :as log]))

;; will do this better at some point
(def api {:keyID "<opsec>"
          :vCode "<opsec>"})

(def eve-api-url "https://api.eveonline.com/char/WalletJournal.xml.aspx")

(defn- query-params
  "takes a map of options and converts into a url encoded query string ready for appending"
  [options]
  (->> options
       (keep (fn [opt-pair]
               (let [param (name (first opt-pair))
                     value (second opt-pair)]
                 (if (and param value)
                   (str param "=" value)
                   nil))))
       (str/join "&")))

(defn- tag-contents
  "Gets the a list of xml data structures for all children that match the tag name"
  [xml tag-name]
  (filter #(= tag-name (:tag %))
          (:content xml)))

(defn- str->num
  "converts a string to a number, returning nil on failure"
  [num-str]
  (let [n (read-string num-str)]
    (if (number? n) n nil)))

(defn- data-rows
  "Extracts the row data from the xml, converting to a list of maps of attributes of each row"
  [xml]
  (-> (tag-contents xml :result)
      (first)
      (tag-contents :rowset)
      (first)
      (tag-contents :row)
      (->> (map :attrs))))

(defn get-transactions
  [start num]
  (let [query-url (str eve-api-url "?" (query-params (merge {:fromID start
                                                             :rowCount num}
                                                            api)))]
    (log/info "Querying API: " query-url)
    (-> (slurp query-url)
        (xml/parse-str)
        (data-rows))))

(defn- min-ref-id
  "Gets the smallest refID in the current batch of transactions, use to walk backwards"
  [batch]
  (apply min (map #(str->num (:refID %)) batch)))

(defn get-all-transactions
  "Recursively grab all transactions and merge together"
  []
  (loop [start nil
         transactions '()]
    (let [batch (get-transactions start 2560)
          batch-size (count batch)]
      (log/info "Batch size: " batch-size)
      (log/info "Starting refd: " start)
      (if (= 0 batch-size)
        transactions
        (recur (min-ref-id batch)
               (concat transactions batch))))))

(defn- filter-transactions
  "filters transactions by one or more parameters"
  [filters]
  )

(defn balance-info
  "Takes a list of transactions and returns the sum of all transactions" 
  [filters]
  (->> (get-all-transactions)
       (reduce (fn [acc t]
                 (let [amount (str->num (:amount t))
                       adder-fn (partial + amount)]
                   (as-> acc res
                     (update res :delta adder-fn)
                     (if (> amount 0)
                       (update res :totalIncome adder-fn)
                       (update res :totalExpense adder-fn)))))
               {:delta 0
                :totalIncome 0
                :totalExpense 0})))

(defn stations
  "Takes a list of transactions and returns a map of stations and number transactions at each"
  []
  (reduce (fn [stations t]
            (let [station (:argName1 t)]
              (if (string? station)
                (update stations 
                        station
                        (fnil inc 0))
                stations)))
          {}
          (get-all-transactions)))



src/eve/wallet.clj
(ns eve.wallet
  "eve api related helpers"
  (:require [clojure.data.xml :as xml]
            [clojure.string :as str]
            [clojure.tools.logging :as log]
            [clj-time.core :as t]
            [clj-time.format :as f]))

(def eve-api-url "https://api.eveonline.com/corp/WalletJournal.xml.aspx")

(defn- query-params
  "takes a map of options and converts into a url encoded query string ready for appending"
  [options]
  (->> options
       (keep (fn [opt-pair]
               (let [param (name (first opt-pair))
                     value (second opt-pair)]
                 (if (and param value)
                   (str param "=" value)
                   nil))))
       (str/join "&")))

(defn- tag-contents
  "Gets the a list of xml data structures for all children that match the tag name"
  [xml tag-name]
  (filter #(= tag-name (:tag %))
          (:content xml)))

(defn- str->num
  "converts a string to a number, returning nil on failure"
  [num-str]
  (let [n (read-string num-str)]
    (if (number? n) n nil)))

(defn- data-rows
  "Extracts the row data from the xml, converting to a list of maps of attributes of each row"
  [xml]
  (-> (tag-contents xml :result)
      (first)
      (tag-contents :rowset)
      (first)
      (tag-contents :row)
      (->> (map :attrs))))

(defn get-transactions
  "Get transactions via REST api,
   * api    map representing eve api key. Must contain :keyID and :vCode fields.
   * start  record t start from (used when getting multiple pages), nil for first page.
   * num    number of records to fetch"
  [api start num]
  (let [query-url (str eve-api-url "?" (query-params (merge {:fromID start
                                                             :rowCount num
                                                             :accountKey 1000}
                                                            api)))]
    (log/info "Querying API: " query-url)
    (-> (slurp query-url)
        (xml/parse-str)
        (data-rows))))

(defn- min-ref-id
  "Gets the smallest refID in the current batch of transactions, use to walk backwards"
  [batch]
  (apply min (map #(str->num (:refID %)) batch)))

(defn get-all-transactions
  "Recursively grab all transactions and merge together"
  [api]
  (loop [start nil
         transactions '()]
    (let [batch (get-transactions api start 2560)
          batch-size (count batch)]
      (log/info "Batch size: " batch-size)
      (log/info "Starting refd: " start)
      (if (= 0 batch-size)
        transactions
        (recur (min-ref-id batch)
               (concat transactions batch))))))


;;
;; {:refTypeID (partial = 2)
;;  :
;;
;;
;;

(defn filter-transaction
  [filters transaction]
  (reduce-kv (fn [match filter condition]
               (and match
                    (if condition
                      (condition (get transaction filter))
                      true)))
             true
             filters))

(defn filter-transactions
  "filters transactions by one or more parameters"
  [filters transactions]
  (->> transactions
       (filter (partial filter-transaction filters))))

(defn balance-info
  "Takes a list of transactions and returns the sum of all transactions

  * Filters consist of 2 parts
    * a keyname matching the keyname in a transaction (xml attribute name from api)
    * a predicate function evaluat=ing whether a transaction shoulkd be kept or removed

  Example:
  {:keyTypeID (partial = \"85\") ;; matches Bountry Prizes type
   :date (fn [date]
           (t/after? (f/parse date) (t/minus (t/now) (t/weeks 1))))}" 
  [api filters]
  (let [transactions (get-all-transactions api)]
    (->> transactions
         (filter-transactions filters)
         (reduce (fn [acc t]
                   (let [amount (str->num (:amount t))
                         adder-fn (partial + amount)]
                     (as-> acc res
                       (update res :delta adder-fn)
                       (if (> amount 0)
                         (update res :totalIncome adder-fn)
                         (update res :totalExpense adder-fn))
                       (update res :count inc))))
                 {:delta 0
                  :totalIncome 0
                  :totalExpense 0
                  :count 0}))))


